File: apps/kotlin/src/main/kotlin/chapter09/PolynomialFeatures.kt
package chapter09

import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.add
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf

internal fun AnyFrame.doubles(column: String): List<Double> = this[column].values().map { (it as Number).toDouble() }

internal fun AnyFrame.selectColumns(names: List<String>): AnyFrame = dataFrameOf(names.map { this[it] })

fun polynomialFeatures(
    df: AnyFrame,
    columns: List<String>,
): AnyFrame =
    pairsWithReplacement(columns).fold(df.selectColumns(columns)) { features, (left, right) ->
        val leftValues = df.doubles(left)
        val rightValues = df.doubles(right)
        features.add(termName(left, right)) { leftValues[index()] * rightValues[index()] }
    }

fun <T> pairsWithReplacement(items: List<T>): List<Pair<T, T>> =
    items.indices.flatMap { i -> (i until items.size).map { j -> items[i] to items[j] } }

fun termName(
    left: String,
    right: String,
): String = if (left == right) "$left^2" else "$left $right"

File: apps/node/src/chapter09/polynomial-features.ts
export function polynomialFeatures<K extends string>(
  rows: readonly Record<K, number>[],
  columns: readonly K[],
): Record<string, number>[] {
  return rows.map((row) => {
    const features: Record<string, number> = {};
    for (const column of columns) {
      features[column] = row[column];
    }
    for (const [left, right] of pairsWithReplacement(columns)) {
      features[termName(left, right)] = row[left] * row[right];
    }
    return features;
  });
}

export function pairsWithReplacement<T>(items: readonly T[]): [T, T][] {
  return items.flatMap((left, i) =>
    items.slice(i).map((right): [T, T] => [left, right]),
  );
}

export function termName(left: string, right: string): string {
  return left === right ? `${left}^2` : `${left} ${right}`;
}
